### ISSUE TITLE
Themis crashes if RSA keys are used incorrectly with Secure Message

### ISSUE THREAD

**Issue by ilammy** [cossacklabs/themis#332]
### Description

Exactly what it says on the tin: preventable programmer error may crash the application.

If the programmer accidentally swaps private and public RSA keys when unwrapping a Secure Message then application may crash.

I've encountered this weird crash when writing tests for error handling of Secure Message. I misplaced the keys in arguments (admittedly, the list _is_ long). I expected a failure, but not a _spectacular_ one! 🦄 

### Environment

**OS:** reproduced on Ubuntu 18.04 as well as on macOS 10.12 (both x86_64)

**Themis version:** 0.10.0 release

**Crypto provider:**
- Ubuntu: OpenSSL 1.1.0g
- Ubuntu: OpenSSL 1.0.2n
- macOS: OpenSSL 1.0.2p
- macOS: LibreSSL 2.8.2

Seems to be _not actual_ for BoringSSL (see notes below).

### Details

Steps to reproduce:

1. Use [this code](https://gist.github.com/ilammy/2510dd8e78ff684e781497e444fccb04#file-crash-c).

Note that `themis_secure_message_unwrap()` is called incorrectly: the private key should go first.

2. Compile and run it:

```console
$ cc crash.c -lthemis
```

**Expected result:**

```console
$ ./a.out
themis_secure_message_unwrap() failed: 12
```

I'd expect Themis to detect the incorrect key usage and gracefully return an appropriate error code (invalid parameter).

**Actual result:**

```console
$ ./a.out
Segmentation fault: 11
```

Hello darkness, my old friend!

### Notes

The issue is not actual for ECDSA keys. If you replace `themis_gen_rsa_key_pair()` with `themis_gen_ec_key_pair()` then the program behaves as expected: Themis returns an error.

The issue is also not actual for BoringSSL (https://github.com/google/boringssl/commit/c8e0f90f83b9ec38ea833deb86b5a41360b62b6a). Unwrapping fails with a different error (11 = THEMIS_FAIL), but the application does not crash. Well, at least on my box.

**Comment by secumod** [comment]
I tried to replicate it in Themis test suite with the following diff
```
@@ -303,7 +303,7 @@ static void secure_message_api_test(void)
        testsuite_fail_unless(THEMIS_INVALID_PARAMETER == themis_secure_message_unwrap(peer_priv, peer_priv_length, NULL, pub_length, ciphertext, ciphertext_length, decryptext, &de
cryptext_length), "themis_secure_message_unwrap: treating encrypted message as signed (NULL peer public key)");
        testsuite_fail_unless(THEMIS_INVALID_PARAMETER == themis_secure_message_unwrap(peer_priv, peer_priv_length, pub, 0, ciphertext, ciphertext_length, decryptext, &decryptext_l
ength), "themis_secure_message_unwrap: treating encrypted message as signed (zero peer public key length)");

-       res = themis_secure_message_unwrap(peer_priv, peer_priv_length, pub, pub_length, ciphertext, ciphertext_length, decryptext, &decryptext_length);
+       res = themis_secure_message_unwrap(pub, pub_length,peer_priv, peer_priv_length, ciphertext, ciphertext_length, decryptext, &decryptext_length);
        if (THEMIS_SUCCESS != res)
        {
                testsuite_fail_if(true, "themis_secure_message_unwrap fail");
```

but the test program did not crash - just returned error as expected. Will try the test code instead...

**Comment by secumod** [comment]
The test code indeed crashes.

**Comment by secumod** [comment]
Traced it to this place: https://github.com/cossacklabs/themis/blob/b98d18b2d7cd99cf04c56c2fed750e6ee5e8ca84/src/soter/openssl/soter_asym_cipher.c#L273

Apparently, some OpenSSL versions do not handle gracefully, if you pass them an `RSA` key without the private part for decryption.

**Comment by secumod** [comment]
Should be fixed by #334
